runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/ReferenceCountingFlinkExecutableStageContextFactory.java

```diff
@@ -115,6 +115,7 @@ private void scheduleRelease(JobInfo jobInfo) {
     int environmentCacheTTLMillis =
         pipelineOptions.as(PortablePipelineOptions.class).getEnvironmentCacheMillis();
     if (environmentCacheTTLMillis > 0) {
+      // Do immediate cleanup if this class is not loaded on Flink parent classloader.
       if (this.getClass().getClassLoader() != ExecutionEnvironment.class.getClassLoader()) {
```

Comment:
@mxm did you verify that this works? If so, then +1

Reply:
No tested but should work since shipped code is loaded via the `FlinkUserCodeClassLoader`. Perhaps we should add a test case for this? I feel like this could easily break.

Reply:
Let's continue investigating this as a follow-up, since it isn't directly linked to this PR. Merging..

Reply:
*Flink* classes (org.apache.flink.*) are always loaded through the parent classloader. The loaded class are always tight to the classloader which was used to load them.

All other classes are first loaded through the child classloader, then the parent by default.

```yaml
classloader.resolve-order: Whether Flink should use a child-first ClassLoader when loading user-code classes or a parent-first ClassLoader. Can be one of parent-first or child-first. (default: child-first)
```
https://ci.apache.org/projects/flink/flink-docs-release-1.5/ops/config.html#common-options

Reply:
`FlinkUserCodeClassLoader` has been removed and the new ones, `ChildFirstClassLoader` and `ParentFirstClassloader` are package-private. So not really an option anyore to check for them.